feat(cdk): Expose props as read/write properties in CFN resources

moofish32 opened the pull request:
This description will be updated before merging: 

This is an initial pass only fixing the generation of property class
members.

There are a good number of collisions with attributes (e.g. Fn::GetAtt)
with the names of the properties. In this implementation I chose to
append `Prop` to all property class members. I'm open to other
suggestions, but my thinkig was attributes have precedence on this name
and the use of L1 property manipulation should be less common.

I have not removed the propertyOverrides feature yet. I am seeking
feedback on the approach I took for minimal impact on generation. Please
see my code comments below.

WIP: #2100

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -138,13 +138,24 @@ export default class CodeGenerator {
       const propSpec = propertiesSpec[propName];
       const additionalDocs = resource.specName!.relativeName(propName).fqn;
-      const newName = this.emitProperty(resource, propName, propSpec, quoteCode(additionalDocs));
+      const newName = this.emitProperty({
```
I'm not sure why I left this an object, I'm going to move back to normal args here.

eladb replied:
I like the props. Definitely more readable.

eladb replied:
Please resolve all conversations that you attended to

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -243,19 +254,11 @@ export default class CodeGenerator {
-    if (tagEnum !== `${TAG_TYPE}.NotTaggable`) {
-      this.code.line();
-      this.code.line('/**');
-      this.code.line(' * The `TagManager` handles setting, removing and formatting tags');
```
Moved this into the emitProps call below. This comment is also removed which I think works. Should I bring this information back and update it because it is already out of date (which is my chief complaint with these type of comments).

eladb replied:
Totally find getting rid of the comment

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -514,15 +520,26 @@ export default class CodeGenerator {
-  private emitProperty(context: genspec.CodeName, propName: string, spec: schema.Property, additionalDocs: string): string {
-    const question = spec.Required ? '' : '?';
-    const javascriptPropertyName = genspec.cloudFormationToScriptName(propName);
+  private emitProperty(props: EmitPropertyProps): string {
```
this is the bulk of m code approach question. I have taken the least change path in this design. However, I debated a larger refactor here where we pull out the concept of this property set that is associated to a resource into an object. Then we would call `renderInterfaceProps()` and `renderClassProps()`. This is a much larger refactor because of where these code generation activities are called and connected. Specifically the validators and mappers are coupled to the current design and would also need refactored.

So I'm seeking feedback on if the investment in a better OO model design here is worth the effort?

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -555,7 +572,13 @@ export default class CodeGenerator {
         const propSpec = propTypeSpec.Properties[propName];
         const additionalDocs = quoteCode(`${typeName.fqn}.${propName}`);
-        const newName = this.emitProperty(resourceContext, propName, propSpec, additionalDocs);
+        const newName = this.emitProperty({
```
this highlights the coupling I'm talking about in my larger comment where I'm seeking feedback.

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -702,3 +725,33 @@ function tagType(resource: schema.ResourceType): string {
+  Class = 'CLASS',
+}
+
+interface EmitPropertyProps {
```
this is dying

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -702,3 +725,33 @@ function tagType(resource: schema.ResourceType): string {
+  container: Container;
+}
+
+// remove export
```
this will be removed

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -702,3 +725,33 @@ function tagType(resource: schema.ResourceType): string {
   return `${TAG_TYPE}.NotTaggable`;
 }
+
+enum Container {
```
I don't love my naming here, what is the right CS term here? Scope? but we use scope in another context.

eladb commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -514,15 +520,26 @@ export default class CodeGenerator {
+      if (props.propName === 'Tags' && schema.isTagProperty(props.spec)) {
+        this.code.line(`public readonly tags: ${TAG_MANAGER};`);
+      } else {
+      this.code.line(`public ${javascriptPropertyName}Prop${line}`);
```
Can you share some examples of conflicts that required adding the `Prop` suffix. I actually think the prop suffix is not that bad, but I am a bit surprised because I was under the assumption that attributes all start with the type name (e.g. `topicName` and props usually don't). But obvsiouly I am wrong.

moofish32 replied:
@eladb -- great follow-up because I didn't look close enough. The collision that I'm seeing with IAM module is actually caused by our `Ref` implementation.

```ts
export class CfnUser extends cdk.CfnResource {
    /**
     * The CloudFormation resource type name for this resource class.
     */
    public static readonly resourceTypeName = "AWS::IAM::User";

    /**
     * @cloudformationAttribute Arn
     */
    public readonly userArn: string;
    /* this is the most common collision */
    public readonly userName: string;

    /**
     * `AWS::IAM::User.Groups`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-groups
     */
    public groups: string[];

    /**
     * `AWS::IAM::User.LoginProfile`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-loginprofile
     */
    public loginProfile: CfnUser.LoginProfileProperty | cdk.Token;

    /**
     * `AWS::IAM::User.ManagedPolicyArns`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-managepolicyarns
     */
    public managedPolicyArns: string[];

    /**
     * `AWS::IAM::User.Path`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-path
     */
    public path: string;

    /**
     * `AWS::IAM::User.PermissionsBoundary`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-permissionsboundary
     */
    public permissionsBoundary: string;

    /**
     * `AWS::IAM::User.Policies`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-policies
     */
    public policies: Array<CfnUser.PolicyProperty | cdk.Token> | cdk.Token;

    /**
     * `AWS::IAM::User.UserName`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-user.html#cfn-iam-user-username
     */
    public userName: string;

    /**
     * Create a new `AWS::IAM::User`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope: cdk.Construct, id: string, props?: CfnUserProps) {
        super(scope, id, { type: CfnUser.resourceTypeName, properties: props });
        this.userArn = this.getAtt('Arn').toString();
        this.userName = this.ref.toString();
    }
 /* ... snip ... */
```

So the issue is actually the `Ref` that has the same name as a property. Would you prefer we suffix these special cases?

eladb replied:
I actually wanted to completely get rid of these RefKinds (see #2345). Happy to include this in this CR if you are inclined, although it will have a pretty substantial breaking effect on the entire framework (not too hard to fix, but annoying).

moofish32 replied:
Can you give me an example of what eliminating RefKinds would look like in the above example? #2345 leads me to believe also eliminating the patching of the spec?

eladb replied:
Yes, basically remove the patching of the spec and the logic that generates bucketName in the above example. See my general comment below.

eladb commented:
Makes me concerned now that mixing up these changes is going to fail miserably and will potentially really confuse people. `userName` is a config while `userArn` is an attribute and `ref` is actually the user name attribute.

@rix0rrr @RomainMuller @skinny85 what do you guys think? 

1. I still think we should get rid of RefKinds for the reasons described in #2345, so basically `bucketName` becomes `ref`.
2. I still think we should expose props as top-level mutable properties in CFN resources.
3. We need to differentiate between *attribute* properties and *prop* properties.

Here's my proposal: let's add a postfix `Attr` to all attribute properties, so it's clear those are deploy-time values (tokens), and also replace the RefKind with a generic `xxxRefAttr`. So we'll have `bucketArnAttr`, `bucketRefAttr`.

Do you think we should maybe remove the bucket prefix (`bucket.arnAttr`, `bucket.refAttr`?).

moofish32 commented:
> Here's my proposal: let's add a postfix Attr to all attribute properties, so it's clear those are deploy-time values (tokens), and also replace the RefKind with a generic xxxRefAttr. So we'll have bucketArnAttr, bucketRefAttr.

To make sure I understand, specifically `bucketRefAttr` no longer has name because the only way we knew it was name was from `RefKind` patch in the specification. 

Also I suppose CFN updating the spec to include this information is off the table?

eladb commented:
Yes, you are correct.

> Also I suppose CFN updating the spec to include this information is off the table?

Not at the moment.

skinny85 commented:
> Here's my proposal: let's add a postfix `Attr` to all attribute properties, so it's clear those are deploy-time values (tokens), and also replace the RefKind with a generic `xxxRefAttr`. So we'll have `bucketArnAttr`, `bucketRefAttr`.

I like that.

> Do you think we should maybe remove the bucket prefix (`bucket.arnAttr`, `bucket.refAttr`?).

I agree with that as well.

moofish32 commented:
So now that I get into this fix I'm starting to think this fix is like drawing the short straw ... this is going to break a lot of things requiring a lot name changes. I think I have a couple of ideas on some ways to shortcut that, but this one could be ugly.

So @rix0rrr @skinny85 @eladb @RomainMuller let's get agreement on the proposal above

> Here's my proposal: let's add a postfix Attr to all attribute properties, so it's clear those are deploy-time values (tokens), and also replace the RefKind with a generic xxxRefAttr. So we'll have bucketArnAttr, bucketRefAttr.

> Do you think we should maybe remove the bucket prefix (bucket.arnAttr, bucket.refAttr?).

This is what role would generate like under that change. Everything referencing `role.roleId` is going to need an update. **_Before I start fixing everything do we agree that we like this or do we want to change it?_**

```ts
export class CfnRole extends cdk.CfnResource {
    /**
     * The CloudFormation resource type name for this resource class.
     */
    public static readonly resourceTypeName = "AWS::IAM::Role";

    /**
     * @cloudformationAttribute Arn
     */
    public readonly arnAttr: string;

    /**
     * @cloudformationAttribute RoleId
     */
    public readonly roleIdAttr: string;

    /**
     * @cloudformationReference Ref
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html
     */
    public readonly refAttr: string;

    /**
     * `AWS::IAM::Role.AssumeRolePolicyDocument`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-assumerolepolicydocument
     */
    public assumeRolePolicyDocument: object | cdk.Token;

    /**
     * `AWS::IAM::Role.ManagedPolicyArns`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-managepolicyarns
     */
    public managedPolicyArns: string[];

    /**
     * `AWS::IAM::Role.MaxSessionDuration`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-maxsessionduration
     */
    public maxSessionDuration: number | cdk.Token;

    /**
     * `AWS::IAM::Role.Path`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-path
     */
    public path: string;

    /**
     * `AWS::IAM::Role.PermissionsBoundary`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-permissionsboundary
     */
    public permissionsBoundary: string;

    /**
     * `AWS::IAM::Role.Policies`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-policies
     */
    public policies: Array<CfnRole.PolicyProperty | cdk.Token> | cdk.Token;

    /**
     * `AWS::IAM::Role.RoleName`
     * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-role.html#cfn-iam-role-rolename
     */
    public roleName: string;

    /**
     * Create a new `AWS::IAM::Role`.
     *
     * @param scope - scope in which this resource is defined
     * @param id    - scoped id of the resource
     * @param props - resource properties
     */
    constructor(scope: cdk.Construct, id: string, props: CfnRoleProps) {
        super(scope, id, { type: CfnRole.resourceTypeName, properties: props });
        cdk.requireProperty(props, 'assumeRolePolicyDocument', this);
        this.arnAttr = this.getAtt('Arn').toString();
        this.roleIdAttr = this.getAtt('RoleId').toString();
        this.refAttr = this.ref.toString();
    }
// ....
```

eladb commented:
Looks good to me, and indeed you are correct that this is going to be break all of our L2 code.

I am wondering if it makes sense to use a common prefix for all attributes instead of a common suffix:

```ts
role.attrRef
role.attrRoleId
role.attrArn
```

I think this will be more organized and easier to discover, and also clearly distinguishes all attributes from props.

Alternatively, we could also just leave the type prefix (`"roleArnAttr"`) and just append `attr`. Both options are okay in my view.

@skinny85 @RomainMuller @rix0rrr any preferences? 

- Option A: `roleRefAttr`, `roleArnAttr`,`roleIdAttr`
- Option B: `attrRef`, `attrArn`, `attrRoleId` (notice the inconsistency between RoleID and ARN because the CFN attribute names are inconsistent in that manner).
- Option C: `roleRefAttr`, `roleArnAttr`, `roleIdAttr`.

My vote I think is option B as it is the most true to CFN and removes all heuristics from our generator, but I don't have a hugely strong opinion - I am fine with any of these options.

moofish32 commented:
I think I like B as well. The prefix organizes better in my mind and maps directly to CFN (even it I don't agree with it, it's been around a long time).

rix0rrr commented:
If I let go how ugly everything is, I suppose I prefer B as well.

moofish32 commented:
@eladb I might have missed this, do we want to "remap" the values in the L2 or make them match? 

```
error: resources must represent all attributes as properties [resource-attribute:@aws-cdk/aws-iam.IRole.attrArn]
```

Right now I was keeping the L2 as `roleArn`, but this means I have to do some flips in the awslint code and that made me wonder if wanted the L2s to match directly?

moofish32 commented:
I pushed an example change of the renaming for aws-iam. Yes, I'm aware more codegen work will be necessary for the `renderProperties( ... )` function etc, but the naming seems like most likely re-work area.

eladb commented on packages/@aws-cdk/aws-iam/lib/lazy-role.ts:
```diff
@@ -76,12 +76,12 @@ export class LazyRole extends cdk.Construct implements IRole {
    */
-  public get roleArn(): string {
-    return this.instantiate().roleArn;
+  public get attrArn(): string {
```
No... we don't want to rename the L2s, just the L1s. The L2s should stay `roleArn`.

eladb commented on tools/awslint/lib/rules/resource.ts:
```diff
@@ -112,7 +112,11 @@ resourceLinter.add({
     const resourceAttributes = new Array<reflect.Property>();
     for (const attr of e.ctx.resource.attributes) {
       const attribute: reflect.Property | undefined = e.ctx.resourceInterface.ownProperties.find(p => p.name === attr);
+      // tslint:disable-next-line:no-console
```
revert

eladb commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -249,21 +249,7 @@ export default class CodeGenerator {
       attributes.push(refAttr);
     }
 
-    // //
```
remove all the refkind JSON patches as well

moofish32 commented:
@eladb @rix0rrr  -- looking for a little guidance here:

I think JSII reflect has an issue where the base class hides the interface that holds the attributes (cfn attributes/refs). 

See here: https://github.com/awslabs/aws-cdk/blob/master/packages/@aws-cdk/aws-iam/lib/group.ts#L54

Versus Role that does a lot of code duplicate here: https://github.com/awslabs/aws-cdk/blob/master/packages/@aws-cdk/aws-iam/lib/role.ts#L112

GroupBase I think is only used to DRY up that import call, but I believe it hides the `@attribute` tags that I've ensured are on the IGroup interface during the aws-lint. It's hard for me to be certain on the details in JSII reflect from my minimal spelunking, but does this sound like I've hit the problem? If so should I remove GroupBase or focus on solving the problem differently?

moofish32 commented:
I was wrong! The issue is unlike the base class for other resources Group does not export GroupBase. What is the right convention here, export the abstract base class for Group (and create one for role)?

eladb commented:
Preferably we shouldn't export the base class, and jsii will copy the declarations (incl. any interface definitions and docstrings) from the base class to the derived classes. Can you describe the issue you are running into?

moofish32 commented:
The issue I'm hitting is that the `@attribute` docTag is on the interface `IGroup`. The L2 for IAM Group does not directly implement `IGroup` instead it extends `GroupBase` which does. Since the `GroupBase` is not exported reflection does not properly find the properties and specifically the CFN attributes and therefore we fail aws-lint. If we export the base class JSII sees it and reflect works.

moofish32 commented:
I've fixed aws-iam and aws-kms package. I've also changed the design a bit. I'm now creating all the class members for CFN resources and if they are optional I'm enabling ` | undefined`. This enables us to pass the props objects directly to the class members and effectively never return to the props object.

eladb commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -266,6 +253,16 @@ export class CfnResource extends CfnRefElement {
+   * Resources that expose mutable properties should override this function to
+   * collect and return the properties object for this resource.
+   */
+  protected get updatedProperites(): { [key: string]: any } {
```
Makes me wonder, why not just implement this behavior in `renderProperties()`? we already have a hook for derived classes to be able to manipulate the synthesized output, seems like a natural place...

moofish32 replied:
moved and simplified some code on the way

moofish32 replied:
combined, let me know what you think.

eladb replied:
Looks great

eladb commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -514,15 +538,29 @@ export default class CodeGenerator {
+    this.docLink(props.spec.Documentation, props.additionalDocs);
+    const line = `: ${this.findNativeType(props.context, props.spec, props.propName)};`;
 
+    if (props.container === Container.Interface) {
```
TBH not sure there is too much value in making this a single method with two very different code branches. It would be more readable to just implement this as two separate methods....

moofish32 replied:
done - split into two functions.

eladb replied:
👍

eladb commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -514,15 +538,29 @@ export default class CodeGenerator {
+    }
+    if (props.container === Container.Class) {
+      const question = props.spec.Required ? ';' : ' | undefined;';
+      const propType = line.replace(';', question);
```
That's ugly... just call `findNativeType` outside and render `line` twice

moofish32 replied:
done

eladb commented on tools/cfn2ts/lib/genspec.ts:
```diff
@@ -189,38 +190,11 @@ export function attributeDefinition(resourceName: CodeName, attributeName: strin
  */
-export function refAttributeDefinition(resourceName: CodeName, refKind: string): Attribute {
-  const propertyName = codemaker.toCamelCase(descriptiveAttributeName(resourceName, refKind));
+export function refAttributeDefinition(): Attribute {
```
feels like this can be omitted and just hard-coded in the codegen

moofish32 replied:
I think we can just delete this and let `this.ref` work from CfnElement, if we are going with the name change to remove `attr` prefix?

eladb replied:
Yes of course. Totally redundant now

moofish32 commented on tools/awslint/lib/rules/resource.ts:
```diff
@@ -152,7 +153,12 @@ resourceLinter.add({
   message: 'resources must represent all cloudformation attributes as attribute properties. missing property: ',
   eval: e => {
     for (const name of e.ctx.cfn.attributeNames) {
```
So I naively tried to do this, but what do you do for all the situations where attributes and property names collide? For example, `CfnCapacityReservation` `availabiltyZone` is both attribute and property (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-capacityreservation.html#cfn-ec2-capacityreservation-availabilityzone).  No easy answer here with the naming convention from CloudFormation that I see? We can eliminate all `attrRef` and just use `ref`.

eladb replied:
What I mean here is that I still want generated L1 attribute properties to have the “attr” prefix but the array “attributeNames” in awslint should be the canonical attribute names without the prefix so it will be easy to work with in awslint

moofish32 replied:
alright -- I was I little dense there, I think I made the changes you wanted now.

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -308,20 +302,43 @@ export default class CodeGenerator {
+    if (propsType && propMap) {
+      this.code.line();
+      if (optionalProps) {
+        this.code.indent(`if (props !== undefined) {`);
```
done

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -334,14 +351,21 @@ export default class CodeGenerator {
     this.code.line(`return ${genspec.cfnMapperName(propsType).fqn}(properties);`);
     this.code.closeBlock();
+    this.code.line();
+    this.code.openBlock(`protected get updatedProperties(): ${propsType.className}`);
```
yep -- I started to do this and stopped. Actually we can simplify `_toCloudFormation` because of this change too.

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -334,14 +351,21 @@ export default class CodeGenerator {
+    this.code.openBlock(`protected get updatedProperties(): ${propsType.className}`);
+    this.code.indent('return {');
+    for (const prop of Object.values(propMap)) {
+      if (prop === 'tags') {
```
I started to do this, but realized tags actually should be in here. The missing part is I was handling tags in the ctor separately and they should just be here. TagManager handles NotTaggable correctly so I simplified the code further (in my opinion).

moofish32 commented on tools/awslint/lib/rules/resource.ts:
```diff
@@ -69,15 +69,16 @@ export class ResourceReflection {
 
-      // an attribute property is a property which starts with the type name
-      // (e.g. "bucketXxx") and/or has an @attribute doc tag.
+      // an attribute property is a property which starts with `attr`
```
still open -- will fix next

moofish32 replied:
Still open - will update next.

moofish32 replied:
updated to only L1 resources

moofish32 commented on tools/cfn2ts/lib/codegen.ts:
```diff
@@ -223,39 +233,23 @@ export default class CodeGenerator {
-    //
-    // Ref attribute
-    //
-    if (spec.RefKind !== schema.SpecialRefKind.None) {
```
still open and will remove next

moofish32 replied:
still open -- will ensure to remove next.

moofish32 replied:
done

moofish32 commented:
Gist of current [kms.generated.ts](https://gist.github.com/moofish32/756853c8d53dc47c91dd321bace9ec17)

moofish32 commented on tools/awslint/lib/rules/resource.ts:
```diff
@@ -152,7 +153,12 @@ resourceLinter.add({
   eval: e => {
     for (const name of e.ctx.cfn.attributeNames) {
-      const found = e.ctx.attributes.find(a => a.names.includes(name));
+      const basename = e.ctx.basename.charAt(0).toLowerCase() + e.ctx.basename.slice(1);
```
I'll come back to fix this assuming we settle on `attr` in the below comments.

moofish32 commented:
@eladb  starting to look at the name collisions: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-flowlog.html#cfn-ec2-flowlog-resourcetype

The resourceType property collides with CfnResource.resourceType. This happens to be the property. We didn't prefix these for readability, but now I'm questioning either a prefix or containing object (again).?

eladb commented:
I prefer not to prefix for these (hopefully rare) edge cases. The base classes should have a pretty minimal API, which I am open to renaming. For example, we can add a “cfn” prefix so it becomes cfnResouceType. What else do we have there?

eladb commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -205,33 +193,21 @@ export class CfnResource extends CfnRefElement {
             DeletionPolicy: capitalizePropertyNames(this, this.options.deletionPolicy),
             Metadata: ignoreEmpty(this.options.metadata),
             Condition: this.options.condition && this.options.condition.logicalId
           }, props => {
```
Curious why we moved renderProperies outside of PostResolve. I suspect the reasoning hasn’t changed or am I missing something?

rix0rrr replied:
Also feel like it should be inside.

moofish32 replied:
moved back, but had to re-separate properties and validator/mapper.

eladb commented on packages/@aws-cdk/cdk/test/test.resource.ts:
```diff
@@ -696,26 +696,52 @@ class Counter extends CfnResource {
-      PROP1: properties.prop1,
-      PROP2: properties.prop2,
-      PROP3: properties.prop3
+  protected get updatedProperites(): { [key: string]: any } {
```
Didn’t we get rid of Updated?

moofish32 replied:
we did -- I'll get that

eladb commented on tools/awslint/lib/rules/cfn-resource.ts:
```diff
@@ -67,13 +66,6 @@ export class CfnResourceReflection {
     this.namespace = fullname.split('::').slice(0, 2).join('::');
 
-    // special-case
-    const basename = this.basename
```
We still need these special cases for rendering the L2 linter

moofish32 replied:
yes - I thought they would come back, but I was in a clean everything out.

eladb commented on tools/awslint/lib/rules/resource.ts:
```diff
@@ -152,7 +153,9 @@ resourceLinter.add({
   eval: e => {
     for (const name of e.ctx.cfn.attributeNames) {
-      const found = e.ctx.attributes.find(a => a.names.includes(name));
+      const lookup = name.startsWith(e.ctx.basename) ?
```
Something here still feels wrong. The purpose of this linter rule is to verify that the L2 has a property for each CFN resource attribute with the format “<type><Attribute>” (I.e bucketArn, topicName,...).

Let’s first render the desired L2 property name in the ResourceReflection class with all the heuristics and special cases (previously those were at the CfnResourceReflection), and then just check that the class has all these properties.

rix0rrr commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -262,8 +238,19 @@ export class CfnResource extends CfnRefElement {
+   * Resources that expose mutable properties should override this function to
+   * collect and return the properties object for this resource.
+   */
+  protected get updatedProperites(): { [key: string]: any } {
```
Typo

moofish32 replied:
removed

moofish32 commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -198,8 +199,8 @@ export class CfnResource extends CfnRefElement {
-            Type: this.resourceType,
-            Properties: ignoreEmpty(this.renderProperties()),
+            Type: this.cfnResourceType,
+            Properties: ignoreEmpty(this.properties),
```
this.properties is now a getter. Having this inside the token forces resolve before renderProperties which is needed for validators that run in the render function. @rix0rrr I could refactor, but now that all render does is validate the resolved properties, this feels ok?

moofish32 commented:
@eladb one more notable change here moving cfn-resource properties to cfnProperties the conflict was in ApiGateway who chose to use properties as a property.

moofish32 commented:
@eladb  - I think I'm pretty close. What are you thoughts on linting? 

Do we want to ensure all L2's expose all the properties or only L1's?

eladb commented:
Indentation [looks off](https://gist.github.com/moofish32/756853c8d53dc47c91dd321bace9ec17#file-kms-generated-ts-L73)

eladb commented:
We already have a way to encode tokens as numbers, so [`| Token`](https://gist.github.com/moofish32/756853c8d53dc47c91dd321bace9ec17#file-kms-generated-ts-L255) is not required for numbers anymore

eladb commented:
This looks good to me. Please rebase and upload the latest version of some generated file so we can validate last time

eladb commented:
Looking forward for resolution of merge conflicts and will merge this once it's ready

moofish32 commented:
Updated generated files:
[Lambda](https://gist.github.com/moofish32/015461b8a7692e53a8411de8d176279f)
[KMS](https://gist.github.com/moofish32/756853c8d53dc47c91dd321bace9ec17)

eladb commented:
Something in the indentation is still off unless it's a copy & paste issue. Can you try to upload the files as-is to somewhere so we can look at the original source?

moofish32 commented:
[generated-examples.tar.gz](https://github.com/awslabs/aws-cdk/files/3298871/generated-examples.tar.gz)

@eladb tarball attached.

moofish32 commented:
proposed commit message

feat(cdk): Expose props in CFN resources and remove propertyOverrides (#2372)

fixes awslabs#2100

Modify the generation of CFN resources to create Class members for each
CloudFormation Property. This removes the need for the property override
solution implemented previously. CloudFormation Resource attributes are
now prefixed with `attr` instead of resource name. The RefKind patches
are removed. All L2 resources have been updated to consume the new
attribute names. The Ref attributes are removed and the new `refAsString`
is used.

Added tagging support for AppSync, AppMesh, StepFunctions.

Token now supports `number` types, so the generated code no longer
treats number as non-tokenizable.

BREAKING CHANGE: All L1 (CFN) Resources have attributes prefixed with
`attr`. For example, in S3 `bucket.bucketArn` is now `bucket.attrArn`.
The property overrides for typed properties has been removed, instead
users can now directly access CloudFormation properties on the class.
